File: backend/app/routes/shares.py
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException, Request, status
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.security import hash_password, verify_password
from app.models.access_log import AccessLog
from app.models.file import File
from app.models.share import ShareLink
from app.models.user import User
from app.routes.dependencies import get_current_user
from app.schemas.share import PublicDownloadRequest, PublicDownloadResponse, ShareCreate, ShareRead
from app.services.cloudinary_service import create_signed_download_url
from app.utils.tokens import create_share_token

router = APIRouter()


def client_ip(request: Request) -> str | None:
    return request.client.host if request.client else None


def write_access_log(
    db: Session,
    action: str,
    request: Request,
    user_id: int | None = None,
    file_id: int | None = None,
) -> None:
    db.add(
        AccessLog(
            user_id=user_id,
            file_id=file_id,
            action=action,
            ip_address=client_ip(request),
        )
    )


def utc_now() -> datetime:
    return datetime.now(timezone.utc)


@router.post("/files/{file_id}", response_model=ShareRead, status_code=status.HTTP_201_CREATED)
def create_share_link(
    file_id: int,
    payload: ShareCreate,
    request: Request,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> ShareLink:
    file_record = db.get(File, file_id)
    if not file_record or file_record.owner_id != current_user.id:
        raise HTTPException(status_code=404, detail="File not found.")

    if payload.expires_at <= utc_now():
        raise HTTPException(status_code=400, detail="Expiry time must be in the future.")

    share_link = ShareLink(
        file_id=file_record.id,
        token=create_share_token(),
        password_hash=hash_password(payload.password) if payload.password else None,
        expires_at=payload.expires_at,
    )
    db.add(share_link)
    db.flush()
    write_access_log(
        db,
        "share_create",
        request,
        user_id=current_user.id,
        file_id=file_record.id,
    )
    db.commit()
    db.refresh(share_link)
    return share_link


@router.post("/{token}/download", response_model=PublicDownloadResponse)
def download_shared_file(
    token: str,
    payload: PublicDownloadRequest,
    request: Request,
    db: Session = Depends(get_db),
) -> PublicDownloadResponse:
    share_link = db.scalar(select(ShareLink).where(ShareLink.token == token))
    if not share_link or share_link.is_revoked or share_link.expires_at <= utc_now():
        raise HTTPException(status_code=404, detail="Share link is invalid or expired.")

    if share_link.password_hash:
        if not payload.password or not verify_password(payload.password, share_link.password_hash):
            raise HTTPException(status_code=401, detail="Invalid share password.")

    file_record = db.get(File, share_link.file_id)
    if not file_record:
        raise HTTPException(status_code=404, detail="File not found.")

    download_url = create_signed_download_url(
        file_record.cloudinary_public_id,
        file_record.cloudinary_resource_type,
        int(share_link.expires_at.timestamp()),
    )
    write_access_log(db, "download", request, file_id=file_record.id)
    db.commit()
    return PublicDownloadResponse(
        file_name=file_record.original_filename,
        download_url=download_url,
    )


@router.post("/{share_id}/revoke", response_model=ShareRead)
def revoke_share_link(
    share_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> ShareLink:
    share_link = db.get(ShareLink, share_id)
    if not share_link:
        raise HTTPException(status_code=404, detail="Share link not found.")

    file_record = db.get(File, share_link.file_id)
    if not file_record or file_record.owner_id != current_user.id:
        raise HTTPException(status_code=404, detail="Share link not found.")

    share_link.is_revoked = True
    db.commit()
    db.refresh(share_link)
    return share_link

File: backend/src/routes/shares.ts
import { Hono } from "hono";
import { zValidator } from "@hono/zod-validator";
import { prisma } from "../lib/prisma";
import { authMiddleware } from "../middleware/auth";
import { User } from "@prisma/client";
import { ShareCreateSchema, PublicDownloadRequestSchema } from "../schemas/share";
import { createShareToken } from "../utils/tokens";
import { hashPassword, verifyPassword } from "../utils/security";
import { createSignedDownloadUrl } from "../services/cloudinary";

export const sharesRouter = new Hono<{ Variables: { user: User } }>();

sharesRouter.post("/files/:file_id", authMiddleware, zValidator("json", ShareCreateSchema), async (c) => {
  const user = c.get("user");
  const fileId = parseInt(c.req.param("file_id"), 10);
  const { expires_at, password } = c.req.valid("json");

  if (isNaN(fileId)) {
    return c.json({ detail: "Invalid file ID." }, 400);
  }

  const fileRecord = await prisma.file.findUnique({ where: { id: fileId } });
  if (!fileRecord || fileRecord.owner_id !== user.id) {
    return c.json({ detail: "File not found." }, 404);
  }

  if (expires_at <= new Date()) {
    return c.json({ detail: "Expiry time must be in the future." }, 400);
  }

  const passwordHash = password ? await hashPassword(password) : null;
  const token = createShareToken();

  const shareLink = await prisma.$transaction(async (tx) => {
    const share = await tx.shareLink.create({
      data: {
        file_id: fileRecord.id,
        token,
        password_hash: passwordHash,
        expires_at,
      },
    });

    await tx.accessLog.create({
      data: {
        user_id: user.id,
        file_id: fileRecord.id,
        action: "share_create",
        ip_address: c.req.header("x-forwarded-for") || null,
      },
    });

    return share;
  });

  return c.json(shareLink, 201);
});

sharesRouter.post("/:token/download", zValidator("json", PublicDownloadRequestSchema), async (c) => {
  const token = c.req.param("token");
  const { password } = c.req.valid("json");

  const shareLink = await prisma.shareLink.findUnique({ where: { token } });
  if (!shareLink || shareLink.is_revoked || shareLink.expires_at <= new Date()) {
    return c.json({ detail: "Share link is invalid or expired." }, 404);
  }

  if (shareLink.password_hash) {
    if (!password) {
      return c.json({ detail: "Invalid share password." }, 401);
    }
    const isValid = await verifyPassword(password, shareLink.password_hash);
    if (!isValid) {
      return c.json({ detail: "Invalid share password." }, 401);
    }
  }

  const fileRecord = await prisma.file.findUnique({ where: { id: shareLink.file_id } });
  if (!fileRecord) {
    return c.json({ detail: "File not found." }, 404);
  }

  const downloadUrl = createSignedDownloadUrl(
    fileRecord.cloudinary_public_id,
    fileRecord.cloudinary_resource_type,
    fileRecord.extension,
    Math.floor(shareLink.expires_at.getTime() / 1000)
  );

  await prisma.accessLog.create({
    data: {
      file_id: fileRecord.id,
      action: "download",
      ip_address: c.req.header("x-forwarded-for") || null,
    },
  });

  return c.json({
    file_name: fileRecord.original_filename,
    download_url: downloadUrl,
  });
});

sharesRouter.post("/:share_id/revoke", authMiddleware, async (c) => {
  const user = c.get("user");
  const shareId = parseInt(c.req.param("share_id"), 10);

  if (isNaN(shareId)) {
    return c.json({ detail: "Invalid share ID." }, 400);
  }

  const shareLink = await prisma.shareLink.findUnique({ where: { id: shareId }, include: { file: true } });
  if (!shareLink || shareLink.file.owner_id !== user.id) {
    return c.json({ detail: "Share link not found." }, 404);
  }

  const updatedShare = await prisma.shareLink.update({
    where: { id: shareId },
    data: { is_revoked: true },
  });

  const { file, ...shareData } = { ...shareLink, is_revoked: true };
  return c.json(shareData);
});
